### PR TITLE
Allowing order changes tweaks

### PR DESCRIPTION
**Pull request by oeoeaio**
Tweaking a couple of things as picked up by @enricostano in #1536

### CODE REVIEW

File: app/controllers/line_items_controller.rb [hunk 1]
```diff
@@ -20,10 +23,12 @@ def bought_items
...
-  # Override default which just redirects
-  # See Spree::BaseController and Spree::Core::ControllerHelpers::Auth
   def unauthorized
-    render text: '', status: 403
+    render nothing: true, status: 401 and return
+  end
+
+  def not_found
+    render nothing: true, status: 404 and return
   end
```
Comment (enricostano) [inline app/controllers/line_items_controller.rb]: if we are going to use `not_found` and `unouthorized` in others API controllers we should probably move them to some OFN base API controller.

File: app/controllers/spree/orders_controller_decorator.rb [hunk 2]
```diff
@@ -224,22 +224,27 @@ def wrap_json_infinity(n)
...
-    order = Spree::Order.complete.find_by_number(params[:id])
-    return order if order.andand.changes_allowed? && can?(:update, order)
-    current_order
+    return @order_to_update if defined? @order_to_update
+    return @order_to_update = current_order unless params[:id]
+    @order_to_update = begin
+      order = Spree::Order.complete.find_by_number(params[:id])
+      order = nil unless order.andand.changes_allowed? && can?(:update, order)
+      order
+    end
```
Comment (enricostano) [inline app/controllers/spree/orders_controller_decorator.rb]: can you move this to another method?

Comment (oeoeaio) [inline app/controllers/spree/orders_controller_decorator.rb]: I can, thought it was on the borderline between readability and excessive breakdown of the method.

File: app/controllers/line_items_controller.rb [hunk 3]
```diff
@@ -1,7 +1,10 @@
 class LineItemsController < BaseController
   respond_to :json
 
-  def index
+  # Taken from Spree::Api::BaseController
+  rescue_from ActiveRecord::RecordNotFound, :with => :not_found
```
Comment (enricostano) [inline app/controllers/line_items_controller.rb]: a `before_filter` would be more explicit IMHO, you probably don't want to rescue all the AR::RNF but only known one

https://github.com/openfoodfoundation/openfoodnetwork/pull/1536#discussion_r116174434

File: app/controllers/line_items_controller.rb (hunk 1 above)
Comment (oeoeaio) [inline app/controllers/line_items_controller.rb]: See my most recent post on discourse. Trying to think about how we structure an API if we're going to build one. Completely take your point, but I'm reluctant to start building something until we have agreement on direction.

File: spec/controllers/line_items_controller_spec.rb [hunk 4]
```diff
@@ -60,8 +61,8 @@
         context "without an order cycle" do
           it "denies deletion" do
             delete :destroy, params
-            expect(response.status).to eq 403
-            expect { item.reload }.to_not raise_error
+            expect(response.status).to eq 401
```
Comment (enricostano) [inline spec/controllers/line_items_controller_spec.rb]: shouldn't this be a `403`? `401` is used for unauthenticated requests. `403` means that even if you're logged in you cannot perform that action.

Comment (oeoeaio) [inline spec/controllers/line_items_controller_spec.rb]: Well that is correct in this instance. This action can never be performed if a current_order_cycle is not present.

**Comment by enricostano** [comment]
Thanks! 💪

I added more comments both here and in the original PR.

File: app/controllers/line_items_controller.rb (hunk 3 above)
Comment (oeoeaio) [inline app/controllers/line_items_controller.rb]: Point taken. As you say, this is a prime example of another convention that should be defined in a lower level controller. Until we make a decision on direction, I am more comfortable implementing something that looks a bit more like a convention than a one-off `before_filter`, but perhaps that is not a very good reason to do it this way in this instance...

Comment (enricostano) [inline app/controllers/line_items_controller.rb]: As you wish, I'm generally used to make clear in the controller what each action need and how that affects the response. Basically: `not_found` implemented in a `Base` controller and `before_filter` in endpoint controller.

File: spec/controllers/line_items_controller_spec.rb (hunk 4 above)
Comment (enricostano) [inline spec/controllers/line_items_controller_spec.rb]: Exactly, the reason of the failure is related to some internal business logic / authorization (`403`) not authentication (`401`).

One thing that could lead to confusion is that in the HTTP status protocol definition the therm "Authorization" is referred to what we refer to "Authentication" in web world 😞 

`CanCan` abilities -> authorization layer `403`
`Devise` and friends -> authentication layer `401`

Or am I missing something regarding the spec? From what I understand a logged in user cannot delete a line item if it is not associated to an order cycle.

I'm sorry if I'm getting this wrong, but at least it's useful to me to understand OFN business logic 🙏

File: app/controllers/spree/orders_controller_decorator.rb (hunk 2 above)
Comment (enricostano) [inline app/controllers/spree/orders_controller_decorator.rb]: I hope we start introducing tools like rubocop and reek and get to an agreement to code style. For me the nested `begin..end` is a clear clue for code split.

One great opportunity that code splitting offers is more documentation surface. That's another thing that would be nice to see in OFN code too 💪

File: spec/controllers/line_items_controller_spec.rb (hunk 4 above)
Comment (oeoeaio) [inline spec/controllers/line_items_controller_spec.rb]: Yep, you're exactly right. My Friday afternoon brain was backwards. Thanks :)

File: app/controllers/spree/orders_controller_decorator.rb (hunk 2 above)
Comment (oeoeaio) [inline app/controllers/spree/orders_controller_decorator.rb]: Yep, agreed. I found it hard to name the new method, so I thought that was a good indicator that the logic didn't need to be split. I've called the new method `changeable_order_from_number` which I'm not al that happy with, but I think the code it a little more readable...

File: app/controllers/line_items_controller.rb (hunk 1 above)
Comment (oeoeaio) [inline app/controllers/line_items_controller.rb]: Ok, so this is the one outstanding question I think. The biggest question for me is how a new hierarchy of json-only controllers fits into our current mix. Do we namespace them separately? If so, what are they called?

**Pull request merged**